### PR TITLE
Dogfood devshells

### PR DESCRIPTION
**Pull request by YorikSar**
Add project.ncl and nickel.lock.ncl, use them for this very flake. For now use only devshells.

Update cachix/install-nix-action to install Nix 2.17 that doesn't evaluate `packages` for all systems on `nix flake check`.

Preparation for dogfooding future improvements from https://github.com/nickel-lang/organist/issues/58

### CODE REVIEW

File: flake.nix [hunk 1]
```diff
@@ -19,7 +19,7 @@
     flake-utils,
     nickel,
   } @ inputs:
-    {
+    builtins.foldl' (prev: next: prev // next prev) {
```
Comment (thufschmitt) [inline flake.nix:22]: This is quite hard to follow. Why can't we instead have something like below?

```nix
let
  outputsFromNickel = ...;
  # whatever else is required ...;
in
(outputsFromNickel ./. inputs
{
  lockFileContents.organist = "./lib/nix.ncl";
}) //
{
  inherit outputsFromNickel;
  # All the pure Nix outputs
}

File: flake.nix [hunk 2]
```diff
@@ -52,67 +52,74 @@
...
-              pkgs.parallel
-              pkgs.gnused
-            ];
-            text = builtins.readFile ./run-test.sh;
-          };
-        in {
-          type = "app";
-          program = pkgs.lib.getExe testScript;
-        };
+            apps =
```
Comment (thufschmitt) [inline flake.nix:79]: Is there a fundamental reason why we can't define the apps in Nickel, or is it just a temporary thing?

**Comment by YorikSar** [comment]
I've added `foldl'` to be able to use things that are not available via `self` without causing infinite recursion like `outputsFromNickel` and `apps` that we need to override to add flake-specific apps. I'm working on a way to define apps in Nickel, so for now we need to keep them in flake.nix. The scope of this PR is to bootstrap dogfooding so that we can migrate more and more bits to Nickel later.

Since `foldl'` approach is confusing to people and doesn't really help on the very next step, I'll remove it and just fall back to bunch of `let`s.

**Review by thufschmitt — APPROVED** [review]
The `let` version is much more readable!

Looks good to go now, let's merge!
